src/ServiceStack/DependencyInjection/DependencyService.cs

```diff
@@ -37,22 +39,82 @@ public DependencyService()
             _containerBuilder = new ContainerBuilder();
         }
 
-        public void RegisterTypeAsItself(Type classType, Sharing sharing = Sharing.None)
+        public void Register(Type implementingType, Sharing sharing, bool registerAsImplementedInterfaces, bool includeNonPublicConstructors)
         {
-            var registration = _containerBuilder.RegisterType(classType);
-            SetSharing(registration, sharing);
+            if (implementingType.IsGenericType)
+            {
+                var registration = _containerBuilder.RegisterGeneric(implementingType);
+                if (registerAsImplementedInterfaces)
+                {
+                    registration = registration.AsImplementedInterfaces();
+                }
+                if (includeNonPublicConstructors)
+                {
+                    registration = registration.FindConstructorsWith(type => type.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic));
+                }
+                registration = SetRegistrationLifetime(registration, sharing);
+            }
+            else
+            {
+                var registration = _containerBuilder.RegisterType(implementingType).AsSelf();
+                if (registerAsImplementedInterfaces)
+                {
+                    registration = registration.AsImplementedInterfaces();
+                }
+                if (includeNonPublicConstructors)
+                {
+                    registration = registration.FindConstructorsWith(type => type.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic));
+                }
+                registration = SetRegistrationLifetime(registration, sharing);
```

Comment:
This section is identical as the one above yes? I would pull them out of the if statement then. The only thing different is the `_containerBuilder.RegisterType` calls